### PR TITLE
Remove liquidity - functional design

### PR DESCRIPTION
**Pull request by marshall-bean**
This PR brings the remove liquidity in line with the functional design.

I also split the quote details into its own component, and refactored Add Liquidity to use it.

### CODE REVIEW

File: projects/dex-ui/src/components/Liquidity/QuoteDetails.tsx [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+        const txEthAmountNumber = formatEther(txEthAmount);
+        const usd = parseFloat(ethPrice.toHuman()) * parseFloat(txEthAmountNumber);
+        setGasFeeUsd(
+          `~${usd.toLocaleString("en-US", {
+            style: "currency",
+            currency: "USD"
+          })}`
+        );
+      }
+    };
```
Comment (0xalecks) [inline projects/dex-ui/src/components/Liquidity/QuoteDetails.tsx:56]: can we extract this to utils, we'll need gas estimation for swap too

Comment (0xalecks) [reply projects/dex-ui/src/components/Liquidity/QuoteDetails.tsx:56]: Also, line 47 and 48... can't you do all that through TokenValue? We can keep USD as a tokenvalue with 2 decimals. not 100% on this one, would need to try it out

File: projects/dex-ui/src/components/Liquidity/RemoveLiquidity.tsx [hunk 2]
```diff
@@ -25,28 +32,35 @@ export const RemoveLiquidity = ({ well, txnCompleteCallback }: RemoveLiquidityPr
...
-  }, [well.tokens]);
+  // useEffect(() => {
+  //   if (well.tokens) {
+  //     const initialAmounts: LiquidityAmounts = {};
+  //     for (let i = 0; i < well.tokens.length; i++) {
+  //       initialAmounts[i] = undefined;
+  //     }
+  //     setAmounts(initialAmounts);
+  //   }
+  // }, [well.tokens]);
```
Comment (0xalecks) [inline projects/dex-ui/src/components/Liquidity/RemoveLiquidity.tsx:63]: Comments

File: projects/dex-ui/src/components/Liquidity/RemoveLiquidity.tsx [hunk 3]
```diff
@@ -71,21 +98,35 @@ export const RemoveLiquidity = ({ well, txnCompleteCallback }: RemoveLiquidityPr
...
         }
-        removeLiquidityTxn = await well.removeLiquidityOneToken(lpTokenAmount, well.tokens![singleTokenIndex], oneTokenQuote, address);
+        const quoteAmountLessSlippage = oneTokenQuote.quote.subSlippage(slippage);
+        removeLiquidityTxn = await well.removeLiquidityOneToken(
+          lpTokenAmount,
+          well.tokens![singleTokenIndex],
+          quoteAmountLessSlippage,
+          address,
+          undefined,
+          { gasLimit: 1000000 }
```
Comment (0xalecks) [inline projects/dex-ui/src/components/Liquidity/RemoveLiquidity.tsx:108]: Not sure if we should be setting gasLimit like this here (and a few below too)

File: projects/dex-ui/src/wells/useLiquidityQuote.tsx [hunk 4]
```diff
@@ -1,65 +1,125 @@
...
+      const nonZeroValues = Object.values(amounts).filter((amount) => amount.value.gt("0")).length;
+      
+      return nonZeroValues === well.tokens?.length;
+    },
+    [amounts, well.tokens]
+  );
+
+  Log.module("useLiquidityQuote").debug("amounts:", amounts);
+  Log.module("useLiquidityQuote").debug("bothAmountsNonZero:", bothAmountsNonZero);
+  Log.module("useLiquidityQuote").debug("removeLiquidityMode:", removeLiquidityMode);
```
Comment (0xalecks) [inline projects/dex-ui/src/wells/useLiquidityQuote.tsx:39]: Can we combine this into one statement, with a more helpful message and an object containing all these props? ex:
```suggestion
  Log.module("useLiquidityQuote").debug("Quote details:", { amounts, bothAmountsNonZero, removeLiquidityMode });
```

**Review by 0xalecks — CHANGES_REQUESTED** [review]

File: projects/dex-ui/src/components/Liquidity/AddLiquidity.tsx [hunk 5]
```diff
@@ -85,16 +98,20 @@ export const AddLiquidity = ({ well, txnCompleteCallback, slippage, slippageSett
   const allTokensHaveMinAllowance = useMemo(() => tokenAllowance.filter((a) => a === false).length === 0, [tokenAllowance]);
 
   const { data: quote } = useQuery(["wells", "quote", "addliquidity", address, amounts, allTokensHaveMinAllowance], async () => {
-    if (!atLeastOneAmountNonzero()) {
+    Log.module("AddLiquidity").debug("Getting quote");
+    Log.module("AddLiquidity").debug("amounts", amounts);
+    Log.module("AddLiquidity").debug("bothAmountsNonZero", bothAmountsNonZero);
+    Log.module("AddLiquidity").debug("address", address);
+    Log.module("AddLiquidity").debug("allTokensHaveMinAllowance", allTokensHaveMinAllowance);
```
[the remark this answers is not in the document]
Comment (marshall-bean) [reply projects/dex-ui/src/components/Liquidity/AddLiquidity.tsx:105]: Eh I can just remove them.

File: projects/dex-ui/src/components/Liquidity/AddLiquidity.tsx [hunk 6]
```diff
@@ -106,7 +123,7 @@ export const AddLiquidity = ({ well, txnCompleteCallback, slippage, slippageSett
         estimate
       };
     } catch (error: any) {
-      Log.module("addliquidity").error("Error during quote: ", (error as Error).message);
+      Log.module("AddLiquidity").error("Error during quote: ", (error as Error).message);
```
[the remark this answers is not in the document]
Comment (marshall-bean) [reply projects/dex-ui/src/components/Liquidity/AddLiquidity.tsx:126]: I was trying to keep it consistent. But I can revert to lower case.

File: projects/dex-ui/src/components/Liquidity/RemoveLiquidity.tsx (hunk 2 above)
Comment (marshall-bean) [reply projects/dex-ui/src/components/Liquidity/RemoveLiquidity.tsx:63]: 🤦

File: projects/dex-ui/src/components/Liquidity/RemoveLiquidity.tsx (hunk 3 above)
Comment (marshall-bean) [reply projects/dex-ui/src/components/Liquidity/RemoveLiquidity.tsx:108]: Well the thing is, I was getting failures without adding a higher gas limit.

File: projects/dex-ui/src/components/Liquidity/QuoteDetails.tsx (hunk 1 above)
Comment (marshall-bean) [reply projects/dex-ui/src/components/Liquidity/QuoteDetails.tsx:56]: 🤔 `TokenValue.mul` does not allow BigNumbers, which we get back from `feeData.maxFeePerGas`.

**Review by 0xalecks — APPROVED** [review]

**Pull request merged**